Sources/App/Commands/VersionReAnalyzer.swift

```diff
@@ -0,0 +1,236 @@
+///   - packages: packages to be analysed
+/// - Returns: future
+func reAnalyzeVersions(client: Client,
+                       database: Database,
+                       logger: Logger,
+                       threadPool: NIOThreadPool,
+                       before cutOffDate: Date,
+                       limit: Int) -> EventLoopFuture<Void> {
+    Package.fetchReAnalysisCandidates(database,
+                                      before: cutOffDate,
+                                      limit: limit)
+        .flatMap { reAnalyzeVersions(client: client,
+                                     database: database,
+                                     logger: logger,
+                                     threadPool: threadPool,
+                                     before: cutOffDate,
+                                     packages: $0) }
+}
+
+
+/// Re-analyze outdated versions for the given list of `Package`s.
+/// - Parameters:
+///   - client: `Client` object
+///   - database: `Database` object
+///   - logger: `Logger` object
+///   - threadPool: `NIOThreadPool` (for running shell commands)
+///   - versionsLastUpdatedBefore: `Date` cut-off for versions to update
+///   - packages: packages to be analysed
+/// - Returns: future
+func reAnalyzeVersions(client: Client,
+                       database: Database,
+                       logger: Logger,
+                       threadPool: NIOThreadPool,
+                       before cutoffDate: Date,
+                       packages: [Package]) -> EventLoopFuture<Void> {
+    // Pick essentials parts of companion function `analyze` and run the for
+    // re-analysis.
+    //
+    // We don't refresh checkouts, because these are being freshed in `analyze`
+    // and would race unnecessarily if we also tried to refresh them here.
+    //
+    // Care should be taken to ensure `reAnalyzeVersions` operates on a
+    // set of versions that is distinct from those `analyze` updates, to
+    // avoid data races.
+    //
+    // Since `reAnalyzeVersions` only updates existing versions, this will be the
+    // case by design, as `analyze` will only add or remove versions, ignoring
+    // existing ones.
+
+    database.transaction { tx in
+        getExistingVersions(client: client,
+                            logger: logger,
+                            threadPool: threadPool,
+                            transaction: tx,
+                            packages: packages,
+                            before: cutoffDate)
+            .flatMap { setUpdatedAt(on: tx, packageVersions: $0) }
+            .flatMap { mergeReleaseInfo(on: tx, packageVersions: $0) }
+            .map { getManifests(logger: logger, packageAndVersions: $0) }
+            .flatMap { updateVersions(on: tx, packageResults: $0) }
+            .flatMap { updateProducts(on: tx, packageResults: $0) }
+            .flatMap { updateTargets(on: tx, packageResults: $0) }
```

Comment:
This is the essential processing pipeline. It effectively shares all logic with the `analysis` pass. There are only a handful of helper functions below which are mainly adapters to make function signatures match.

The only additions are:

- `getExistingVersions` which returns the existing versions instead of the `added, deleted` pair in `analysis`
- `setUpdatedAt` which updates `update_at` for processed versions (analysis only adds versions, so this is not needed there)
- `Package.fetchReAnalysisCandidates` which is the candidate selection specific to re-analysis